src/wallet/wallet.cpp

```diff
@@ -1422,12 +1414,13 @@ void CWallet::BlockConnected(const CBlock& block, const std::vector<CTransaction
     // to abandon a transaction and then have it inadvertently cleared by
     // the notification that the conflicted transaction was evicted.
 
+    m_last_block_processed_height += 1;
```

Comment:
This height accounting in the wallet scares me slightly. I think that prior to this change, the wallet tolerates any bug in the node that sends duplicate `BlockConnected()` or `BlockDisconnected()` calls over the CValidationInterface. After this change, a duplicate `BlockConnected` notification could cause the wallet to get into a very bad inconsistent state.

Is it possible to update the CValidationInterface to also pass the block height when sending BlockConnected/BlockDisconnected notifications?

Reply:
You're right, it's far better to rely on message content than counting the message in itself.
To do, I've extended BlockDisconnected, for BlockConnected we have CBlockIndex it was just masked between Notification and CWallet